### ISSUE TITLE
Callbacks do not run if no column attribute changed

### ISSUE THREAD

**Issue by akadusei** [luckyframework/avram#432]
https://github.com/luckyframework/avram/blob/cce8fee61768b3bc4cbc28fb4f038421a26e1df6/src/avram/save_operation.cr#L300

In `Avram::SaveOperation`s, `after_save` and `after_commit` callbacks do not run if no database column attribute changed. I think they should.

The current situation means that if an update operation (let's call this the child operation) is performed in any of the said callbacks of another update operation (let's call this the parent operation), the child operation is **not** guaranteed to run.

This is because the parent operation may not have any column attributes changed, so none of the said callbacks run, even though the parent operation is marked as `SaveStatus::Saved`.

Possible Use Cases:

- In a `ResetPassword` (a `User::SaveOperation`), one may want the password reset token deleted `after_commit`, whether or not the password actually changed. (Note: In this operation, the password is the only column that can change).

- In a 1:1 association, one may give a parent operation `attribute`s to be passed to a child operation, to be saved by the child in the parent's `after_save` callback. If `#update` is called on the parent but no column attribute of the parent changed, the parent would be marked as `SaveStatus::Saved` but the child operation won't run.

My proposed solution would be to update the `#save` method to:

```crystal
def save : Bool
  if valid? && (!persisted? || changes.any?)
    # ...
  elsif valid? && changes.empty?
    after_save(record.not_nil!)
    after_commit(record.not_nil!)
    self.save_status = SaveStatus::Saved
    true
  else
    # ...
  end
# UPDATE: 2020-07-30
# Mark as failed if a nested operation rolls back a database transaction
rescue Rollback
  mark_as_failed
  false
end
```

One could always check with `#changes` (or similar) to skip (or not) any actions to be performed in any of the said callbacks.

**Comment by jwoertink** [comment]
That's a good point. Once I get back to working on https://github.com/luckyframework/avram/pull/369 I'll be sure to include this.

**Comment by paulcsmith** [comment]
Left a comment here with what I think we can do instead https://github.com/luckyframework/avram/pull/481#issuecomment-713881686

TL;DR I think we should keep the current behavior and add a new set of callbacks for running after a successful run that runs if the operation is valid, even if the record was not saved
